AndroidNativeKotlinTemplate/app/src/main/java/com/salesforce/androidnativekotlintemplate/MainActivity.kt

```diff
@@ -90,21 +93,38 @@ class MainActivity : SalesforceActivity() {
 
     override fun onResume() {
         // Hide everything until we are logged in
-        findViewById<ViewGroup>(R.id.root).visibility = View.INVISIBLE
+        findViewById<ViewGroup>(root).visibility = View.INVISIBLE
 
         // Create list adapter
         listAdapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, ArrayList())
         findViewById<ListView>(R.id.contacts_list).adapter = listAdapter
 
+        // Check for and use the intent's QR code login URL if applicable.
+        val buildRestClientOnResumePreviousValue = buildRestClientOnResume
+        if (isQrCodeLoginUrlIntent(intent ?: return)) {
+            /*
+             * Temporarily request the Salesforce Mobile SDK REST Client not be
+             * built on resume such that the default login activity will not
+             * display and conflict with the login activity for QR code log in.
+             *
+             * If the display of the default login activity was no longer a
+             * side-effect of getting the REST client, this variable could be
+             * removed.
+             */
+            buildRestClientOnResume = false
```

Comment:
Much like a similar comment on https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/2628, the app's ability to receive data in main activity is greatly limited by the automatic display of the default log in activity.  This returns some control to the app until possibly some of that tightly coupled logic can be refactored.

Reply:
I don't think I have a quick alternative but I really don't like this.  This feels like something the SDK should be handling internally

Reply:
Could we bury that flag deeper maybe between SalesforceSDKMgr and SalesforceActivityDelegate ??

Reply:
Maybe QR Code login needs it's own activity?  We don't have this problem with Native Login because that is the activity that gets launched when on resume happens with no users logged in.

Reply:
I’m glad we mentioned handling this “internally to the SDK”. The trouble here is that SalesforceActivity does attempt to handle this internally - with no option to make exceptions.  Before we go the route of creating a “main” activity that doesn’t inherit from SalesforceActivity (A suggestion from @Brandonpage I really considered - it makes sense, though hear me out), I have a thought on how we might avoid some redundant code and maybe make SalesforceActivity a more re-usable class for the future. 

The goal is passing the QR Code Login URL/Frontdoor URL/Code Verifier from MainActivity to our existing LoginActivity so we get gobs of code re-use.  I’m sensing we need a loosely coupled way to do things like this in general, but our big singleton is the opposite of loosely coupled (Ain’t it a stinker?). I’d love to avoid adding more properties to that crazy thing.

The trouble is SalesforceActivity is hard-coded to always start LoginActivity in just one way and the app doesn’t have an API to change that.  `onResume` thinks it has to start LoginActivity every time and since that call is buried in ClientManager a few levels down the stack it’s not customizable.

The funny thing is SalesforceActivityDelegate has a `public void onResume(boolean buildRestClient)` and it’s really the `buildRestClient` the app would like to set to `false`.  That’s been a public method for seven years, but the delegate itself is private.  Can we give that to the apps? We simply need a means for the app to use this existing and amazingly public parameter.  Maybe we just need a great name and doc comment so the semantic makes sense?

For now, is there a way to clean up this new buildRestClientOnResume property so it makes sense, maybe even in a non-deprecated way?

This is just my thinking, but ideal future state is we don’t force a specific LoginActivity in on-resume.  If LoginActivity provides some app-consumable options the app could have the API to start LoginActivity itself.  That’s where we need an API allowing the app to opt-out of the automatic, default LoginActivity - When it wants to.

No after hours reply needed 💤 Thanks for pondering a great team reply on this little stumper.

Reply:
Sleeping on it, I tried out cleaning up the new property this morning so it's benefits and usage are much more clear.

👉🏻 https://github.com/forcedotcom/SalesforceMobileSDK-Templates/pull/427/commits/dd345a384bee8bbe69b2493171e102dd27a28afc

As I thought about my own experience developing QR code log in as an MSDK app developer, I really thought there was something missing without this property.  The concern I see myself having, in the app developer persona, is that I could use this template and SalesforceActivity to bootstrap my new app and all would seem great - at first.  Then, after investing in building on SalesforceActivity I realize later there's no way to customize the activity's strict on-resume flow (The runtime enforces the call to `super.onResume()`).  It'd be tough, then, to have to retrofit off SalesforceActivity and onto an entirely custom one.

Since MSDK already has a property we can present by-proxy to the app developer we can open up SalesforceActivity to a new scale and provide an off-ramp from the hard-coded on-resume flow - only when needed.

Thoughts? We _could_ go back to adding the new properties on SalesforceSDKManager singleton, though I feel we have a nice opportunity to modernize this activity and avoid building on our old pattern.

Reply:
Just for reference on how odd it looked using the singleton variables on SalesforceSDKManager, that's here 👉🏻 https://github.com/forcedotcom/SalesforceMobileSDK-Templates/commit/e5980e4c418bd73e8f1cde5d50b8d0786c4e0957#diff-dfb6fb02b5c24349900f92938b21c03c85aad96aa277a2c36ec2886a51696c80R189